packages/block-editor/src/components/use-setting/index.js

```diff
@@ -130,31 +131,44 @@ export default function useSetting( path ) {
 				),
 			];
 
-			for ( const candidateClientId of candidates ) {
-				const candidateBlockName =
-					select( blockEditorStore ).getBlockName(
-						candidateClientId
-					);
-				if (
-					hasBlockSupport(
-						candidateBlockName,
-						'__experimentalSettings',
-						false
-					)
-				) {
-					const candidateAtts =
-						select( blockEditorStore ).getBlockAttributes(
+			result = applyFilters(
```

Comment:
My understanding is that the goal is to allow consumers to filter the settings almost on a per-block instance. Two things here:

- the only data that provides contextual information about the block instance to be filtered is the "candidates" list, which are the ancestors of that block. If we aim to filter per-instance, what do you think of only passing the `clientId` instead? Gutenberg can't anticipate all data a consumer may need (`blockName`, the ancestors, etc.), so it'd be best to stay flexible and make the filter about its purpose (filter particular block instances).

- if the filter is a short-circuit one: I find it more readable and more future-proof (code after it may change) if we just return earlier.

```js
result = applyFilters( 'blockEditor.useSetting.before', clientId );
if ( undefined !== result ) {
  return result;
}
```

Reply:
> the only data that provides contextual information about the block instance to be filtered is the "candidates" list, which are the ancestors of that block. If we aim to filter per-instance, what do you think of only passing the clientId instead? Gutenberg can't anticipate all data a consumer may need (blockName, the ancestors, etc.), so it'd be best to stay flexible and make the filter about its purpose (filter particular block instances).

We based the code here very closely on [@mcsf’s original suggestion](https://github.com/WordPress/gutenberg/issues/41547#issuecomment-1249590021). We agree that it’s impossible and unnecessary to pass all possible context to the filter here. It would still be very useful for our purposes to include `blockName` - we use this to quickly short-circuit on the filter side if the block doesn’t have any special context associated with it. We can query for other data we need, but it would be great if we could keep the `blockName` available since `useSetting()` gets called quite often and it’s an easy way to return quickly via a fast lookup. Looking around the codebase a bit we found [`editor.Autocomplete.completers` filter](https://github.com/WordPress/gutenberg/blob/e0464ed/packages/block-editor/src/components/autocomplete/index.js#L50-L54) doing something similar, but often filters pass the whole `blockType`. Just the `blockName` would be very helpful for us to make the filter implementation return ASAP when it’s not needed for a block.

> if the filter is a short-circuit one: I find it more readable and more future-proof (code after it may change) if we just return earlier.
> using path vs normalizedPath in the filter.

Both of these make sense. We can push up a change that swaps the `normalizedPath` for the path, and returns earlier from the filter so it acts as a short-circuit one.

Reply:
Absolutely understand that. We shall add `clientId` to the API as well, for extra flexibility.

Reply:
I've removed `candidates` and add `clientId` to the filter instead. It should be simpler now!